tests/drivers/mtd_spi_nor/main.c

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright (C) 2024 Technische Universität Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @brief       Test for the MTD SPI NOR Driver
+ *
+ * @author      Christopher Büchse <christopher.buechse@tuhh.de>
+ */
+
+#include <string.h>
+#include <errno.h>
+
+#include "busy_wait.h"
+#include "embUnit.h"
+
+#include "mtd_spi_nor.h"
+#include "../../drivers/mtd_spi_nor/include/mtd_spi_nor_defines.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+#define ENABLE_TRACE    0
+#define TRACE(...)      DEBUG(__VA_ARGS__)
+
+#define TEST_MTD mtd_dev_get(MTD_NUMOF-1) /* Make sure to get the last device */
+
+/* This function was copied from drivers/mtd_spi_nor/mtd_spi_nor.c */
+static inline spi_t _get_spi(const mtd_spi_nor_t *dev)
+{
+    return dev->params->spi;
+}
+
+/* This function was copied from drivers/mtd_spi_nor/mtd_spi_nor.c */
+static void mtd_spi_cmd(const mtd_spi_nor_t *dev, uint8_t opcode)
+{
+    if (IS_ACTIVE(ENABLE_TRACE)) {
+        TRACE("mtd_spi_cmd: %p, %02x\n",
+              (void *)dev, (unsigned int)opcode);
+    }
+    spi_transfer_byte(_get_spi(dev), dev->params->cs, false, opcode);
+}
+
+/* This function was copied from drivers/mtd_spi_nor/mtd_spi_nor.c */
+static void mtd_spi_cmd_read(const mtd_spi_nor_t *dev, uint8_t opcode, void *dest, uint32_t count)
+{
+    if (IS_ACTIVE(ENABLE_TRACE)) {
+        TRACE("mtd_spi_cmd_read: %p, %02x, %p, %" PRIu32 "\n",
+              (void *)dev, (unsigned int)opcode, dest, count);
+    }
+    spi_transfer_regs(_get_spi(dev), dev->params->cs, opcode, NULL, dest, count);
+}
+
+/* This function was copied from drivers/mtd_spi_nor/mtd_spi_nor.c */
+static void mtd_spi_cmd_write(const mtd_spi_nor_t *dev, uint8_t opcode,
+                              const void *src, uint32_t count)
+{
+    if (IS_ACTIVE(ENABLE_TRACE)) {
+        TRACE("mtd_spi_cmd_write: %p, %02x, %p, %" PRIu32 "\n",
+            (void *)dev, (unsigned int)opcode, src, count);
+    }
+    spi_transfer_regs(_get_spi(dev), dev->params->cs, opcode,
+                      (void *)src, NULL, count);
+}
+
+/* This function was copied from drivers/mtd_spi_nor/mtd_spi_nor.c */
+static inline void wait_for_write_enable_set(const mtd_spi_nor_t *dev)
```

Comment:
Why does the test copy internal driver code?
Do you actually test the driver or the copy of it that lives here?

Reply:
The issue is that I need to access some low level stuff of the Flash to prepare the test condition. Precisely speaking I am setting the block protect bits to trigger the program/erase error.
The only other way to trigger it would be to have a device with faulty memory cells, which @vincent-d had, but I don't.

The actual relevant part of the test is not this low level stuff, but whether or not the mtd_write_page_raw and mtd_erase return the correct value. So the relevant test is just these two functions. The functions I copied are accessed multiple times in the higher level functions anyway, so they are tested implicitly as well.

The reason I copied the functions is that I can not access them (and they usually shouldn't be accessed from the outside). However, just inserting the code into the test case seemed to be redundant and therefore I copied the functions.
If there is a better way, I would love to change it.